### PR TITLE
8316632: Shenandoah: Raise OOME when gc threshold is exceeded

### PR DESCRIPTION
**Pull request by earthling-amzn**
Shenandoah will run back-to-back full GCs and _almost_ grind mutator progress to a halt before eventually exhausting memory. This change will have Shenandoah raise a gc threshold exceeded exception if the collector fails to make progress after `ShenandoahNoProgressThreshold` full GC cycles (default is 3).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316632](https://bugs.openjdk.org/browse/JDK-8316632): Shenandoah: Raise OOME when gc threshold is exceeded (**Enhancement** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - no project role) ⚠️ Review applies to [f68a942b](https://git.openjdk.org/jdk/pull/15852/files/f68a942bfec8accf584602bc0091db1cc86973a5)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - **Reviewer**) ⚠️ Review applies to [f68a942b](https://git.openjdk.org/jdk/pull/15852/files/f68a942bfec8accf584602bc0091db1cc86973a5)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15852/head:pull/15852` \
`$ git checkout pull/15852`

Update a local copy of the PR: \
`$ git checkout pull/15852` \
`$ git pull https://git.openjdk.org/jdk.git pull/15852/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15852`

View PR using the GUI difftool: \
`$ git pr show -t 15852`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15852.diff">https://git.openjdk.org/jdk/pull/15852.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15852#issuecomment-1728523164)

### CODE REVIEW

**Review by kdnilsen — APPROVED** [review]
Looks good to me.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:940]: Move this block to `ShenandoahHeap::allocate_memory`? This would require adding a (default) argument to that method, but it is a good tradeoff for putting all mechanics around this into a single method.

**Comment by shipilev** [comment]
Please also run `tier1 tier2 tier3` with Shenandoah? In `hotspot:tier1`, I am seeing the failure here:

```
TEST: gc/InfiniteList.java

STDERR:
Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded
        at java.base/java.util.jar.Manifest$FastInputStream.<init>(Manifest.java:421)
        at java.base/java.util.jar.Manifest$FastInputStream.<init>(Manifest.java:416)
        at java.base/java.util.jar.Manifest.read(Manifest.java:287)
        at java.base/java.util.jar.Manifest.<init>(Manifest.java:101)
        at java.base/java.util.jar.Manifest.<init>(Manifest.java:88)
        at java.base/java.util.jar.JarFile.getManifestFromReference(JarFile.java:431)
        at java.base/java.util.jar.JarFile.getManifest(JarFile.java:405)
        at java.base/jdk.internal.loader.URLClassPath$JarLoader$2.getManifest(URLClassPath.java:853)
        at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:848)
        at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:760)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:681)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:639)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
        at com.sun.javatest.regtest.agent.MainWrapper.main(MainWrapper.java:94)

```

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:940]: I put it here to keep it on the slow(er) path so that allocations for LABs won't have additional checks on the progress counter. Also, a failed LAB allocation will come back to `mem_allocate` to try for a shared allocation.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:292]: Can we call this `_gc_no_progress_count` instead, so it's consistent with the corresponding `get_` method? The difference in naming is unnecessary at best and potentially confusing for maintenance in the long run, at worst.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp:124]: Re "normal allocation" : what is not a normal (i.e. abnormal?) allocation?

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp:125]: Who is allowed to call the non-blocking version? What is the semantics of block or don't block? Can the documentation be extended to elaborate on its use?

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:947]: I am confused. We have noted that we did more than `ShenandoahNoProgressThreshold` collections at this point. So we will fail the allocation request, set `gc_overhead_limit_was_exceeded`, and return `null` to the requester.

What happens at the caller, in response to this set of conditions?

Why then does it make sense to `notify_gc_progress()`, which clears the counter of `_gc_no_progress_count`? Where was the ostensible progress in this case?

**Review by ysramakrishna — COMMENTED** [review]
I have a few questions; thanks for clearing up my confusion! :-)

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp:124]: I believe 'normal' here is a mutator allocation. I can update this comment.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp:125]: When a mutator allocation fails, it will block on this call and wait to be notified by the control thread to retry the allocation. The notification follows a degenerated or full GC.

Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp:125]: With this change, the mutator is also allowed to call the non-blocking version when it cannot allocate because the GC overhead has been exceeded. I didn't make sense to me to have the mutator block and wait for a GC cycle when the GC threshold is already exceeded.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:947]: Ultimately, the caller will have an OOME raised. The `_gc_no_progress_count` is cleared to allow subsequent mutator allocations to resume the slower block-and-retry loop in `ShenandoahHeap::allocate_memory`.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:292]: Yes, I'll rename this.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:947]: OK, thanks, that makes sense. Can you add a comment line to that effect?

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp:125]: OK, thanks, that makes sense.

**Review by ysramakrishna — APPROVED** [review]
LGTM, modulo a couple of clarifying documentation comments, as discussed.

File: test/hotspot/jtreg/TEST.groups
Comment (shipilev) [inline test/hotspot/jtreg/TEST.groups:612]: Adding new test groups require additional attention. I think this should be removed.

Note that during debugging/testing, you can run the same by:

```
$ make test TEST="runtime/reflect/ReflectOutOfMemoryError.java gc/InfiniteList.java runtime/ClassInitErrors/TestOutOfMemoryDuringInit.java" 
```

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:877]: Can this be moved to the block that already does allocation-after-gc retries? That block already exits with `nullptr` (implies delivering OOME), and it already calls `handle_alloc_failure` (thus triggering GC). We "only" need it to specialize for `is_lab_alloc` and `ShenandoahNoProgressThreshold`?

This PR changes that block anyway...

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:902]: There is `type_string()` that can be used instead of `is_lab_alloc()` here.

**Review by shipilev — COMMENTED** [review]
Apologies, but I have more comments.

File: test/hotspot/jtreg/TEST.groups
Comment (earthling-amzn) [reply test/hotspot/jtreg/TEST.groups:612]: I'll revert this.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp]: Yes, I've made this change as you suggest, but I'm not sure it improves readability. Note that when the `ShenandoahNoProgressThreshold` is exceeded, we do _not_ retry the allocation or wait for another gc cycle to complete.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp]: TIL - thank you.

File: test/jdk/com/sun/jdi/EATests.java
Comment (shipilev) [inline test/jdk/com/sun/jdi/EATests.java:274]: Let's do explicit `ZGCIsSelected` and `ShenandoahIsSelected`. Unfortunately, "Concurrent GC" is fairly ambiguous.

**Review by shipilev — APPROVED** [review]
All right, this looks fine.

**Pull request closed without merging**
